### DISCUSSION TITLE
Is there a way to disable history behavior?

### DISCUSSION THREAD

**Discussion by unphased** [marlonrichert/zsh-autocomplete#424] [Q&A] [answered]
I am used to using fzf's built-in ctrl+R history search. I noticed that by setting it to load after autocomplete that I can get this keybind to function, however various keystrokes like up arrow will bring up a menu provided by autocomplete that is limited to a fixed number of entries. I understand that this limit is configurable, but the mere fact that this config exists and is defaulted to a value of 16 is a strong hint that setting it to a large value will have a performance impact. I could dive in-depth as to why the behavior present in zsh-autocomplete is inferior to that of fzf but this seems unnecessary. 

So I would like to disable both types of history related functionality from this plugin entirely and have up arrow function like it does in vanilla zsh, which seems to be that it will just pull up whatever history whose beginning exactly matches what has been typed in (before fzf I used to use `history-substring-search`, but fzf built in behavior is so far superior that I never set up history-substring-search anymore). 

That is to say, I'd like to use this autocomplete plugin only for the display of completions and none of its history-related features.

Is there a way to achieve this? [config doc](https://github.com/marlonrichert/zsh-autocomplete/blob/main/.zshrc) is not complete enough. I will try to look around in the source and post back if i find suitable configuration that I can set.

EDIT!

I've realized that the history feature of this plugin has some redeemable capabilities!! One thing that I haven't had a good way to do before is to issue multiple commands from before into a single new command. However the history completion menu in this plugin makes that rather easy to do. The plot thickens!

**Comment by unphased** [comment]
I have attempted to put this into my zshrc after the plugin is loaded. 

```
bindkey '\eOA' up-line-or-beginning-search
bindkey '\e[A' up-line-or-beginning-search
bindkey '\eOB' down-line-or-beginning-search
bindkey '\e[B' down-line-or-beginning-search
```
This is a small amount of progress. Because after the up arrow is pressed, autocomplete does interfere with the situation in some kind of way that the "beginning-search" gets clobbered such that the buffer being beginning-searched becomes the full buffer at this point, thus preventing beginning-search from behaving as intended. At the end of the day it is not too bad to re-enter fzf as a relief valve...

**Reply by unphased** [reply]
I am also starting to understand the default value of 16. Basically it will expand to fill space, so if I set the value to 200 or 9999 it will basically fill the entire terminal window with the completion content, which is not ideal since recent past history is getting clobbered by this. This would highlight the need for a separate limit in lines that history gets pulled to allow you to choose from, vs the number of lines that is allotted the completion.  Hope that makes sense... 

Still hoping that there is a way to disable history navigation entirely. I still really want the completion behavior. There is so much to learn from the completions that get shown. I don't want to live without that anymore but the history behavior is making it difficult!

**Comment by marlonrichert** [comment, marked as answer]
> That is to say, I'd like to use this autocomplete plugin only for the display of completions and none of its history-related features.
>
> Is there a way to achieve this? [config doc](https://github.com/marlonrichert/zsh-autocomplete/blob/main/.zshrc) is not complete enough.

The config doc does in fact say how to achieve this:
https://github.com/marlonrichert/zsh-autocomplete/blob/d6a0ff574679280af52258ee6d2dcd3857283270/.zshrc#L67
https://github.com/marlonrichert/zsh-autocomplete/blob/d6a0ff574679280af52258ee6d2dcd3857283270/.zshrc#L73

So, in order to get the default Zsh behavior of the up and down arrows, you need to add this to your `.zshrc` _after_ sourcing Zsh Autocomplete: 
```zsh
bindkey '\e[A' up-line-or-history
bindkey '\eOA' up-line-or-history
bindkey '\e[B' down-line-or-history
bindkey '\eOB' down-line-or-history
```

---

> I am also starting to understand the default value of 16. Basically it will expand to fill space, so if I set the value to 200 or 9999 it will basically fill the entire terminal window with the completion content, which is not ideal since recent past history is getting clobbered by this. This would highlight the need for a separate limit in lines that history gets pulled to allow you to choose from, vs the number of lines that is allotted the completion.

Again, that's already possible and documented.

For up arrow:
https://github.com/marlonrichert/zsh-autocomplete/blob/d6a0ff574679280af52258ee6d2dcd3857283270/.zshrc#L25

For Ctrl-R and Ctrl-S:
https://github.com/marlonrichert/zsh-autocomplete/blob/d6a0ff574679280af52258ee6d2dcd3857283270/.zshrc#L28

**Reply by unphased** [reply]
Thanks. Those key binds did help for me. 

It is possible that I had not done a comprehensive test of the functionality, of `list-lines`. But I found no way to make it only use a max of 16 lines worth of space while allowing 9999 values to be scrollable within that space. I could set the value to 9999 to allow scrolling thru all history, but this makes the menu fill the screen and clobber all of the visible terminal scrollback. (This is just a repeat attempt to describe what I described earlier)

**Reply by marlonrichert** [reply]
> But I found no way to make it only use a max of 16 lines worth of space while allowing 9999 values to be scrollable within that space. 

Because that's currently not possible with Zsh's completion system. If you want that to be possible, you'll need to submit a feature request or patch to the [`zsh-workers` mailing list](https://zsh.sourceforge.io/Arc/mlist.html). 🙂

**Reply by unphased** [reply]
Got it! Really appreciate your sharing your knowledge of zsh and pushing its boundaries!
